### PR TITLE
Namespace configurators

### PR DESCRIPTION
**Pull request by MarinPostma**
In this PR, we rework how the namespace store instantiates namespaces to make a place for libsql wal replicas. To that end, we introduce the `NamespaceConfgurators,` which are responsible for instantiating replicas, primaries, or schema databases. The `NamespaceConfigurators` is set up during program initialization with whatever state is required to instantiate a particular namespace type. Depending on the configuration, the `NamespaceStore` will call the appropriate namespace configurator to load the namespace.

This is useful for libsql-wal since we must keep a handle to the `libsql-wal` registry to get an injector or a replicator instance.

To reduce the scope of this PR, the implementation of the replication protocol using libsql wal will be integrated later.

### CODE REVIEW

File: libsql-server/src/namespace/configurator/primary.rs [hunk 1]
```diff
@@ -4,125 +4,192 @@ use std::{path::Path, pin::Pin, sync::Arc};
...
+        let connection_maker = Arc::new(connection_maker);
+
+        if meta_store_handle.get().shared_schema_name.is_some() {
+            let block_writes = block_writes.clone();
+            let conn = connection_maker.create().await?;
+            tokio::task::spawn_blocking(move || {
+                conn.with_raw(|conn| -> crate::Result<()> {
+                    setup_migration_table(conn)?;
+                    if has_pending_migration_task(conn)? {
+                        block_writes.store(true, Ordering::SeqCst);
```
Comment (LucioFranco) [inline libsql-server/src/namespace/configurator/primary.rs:86]: if this fails do we just keep it blocked forever? Is that going to be a problem?

File: libsql-server/src/namespace/configurator/mod.rs [hunk 2]
```diff
@@ -73,13 +99,30 @@ impl NamespaceConfigurators {
...
-    ) -> Pin<Box<dyn Future<Output = crate::Result<super::Namespace>> + Send + 'a>>;
+    ) -> Pin<Box<dyn Future<Output = crate::Result<Namespace>> + Send + 'a>>;
+
+    fn cleanup<'a>(
+        &'a self,
+        namespace: &'a NamespaceName,
+        db_config: &'a DatabaseConfig,
+        prune_all: bool,
+        bottomless_db_id_init: NamespaceBottomlessDbIdInit,
+    ) -> Pin<Box<dyn Future<Output = crate::Result<()>> + Send + 'a>>;
```
Comment (LucioFranco) [inline libsql-server/src/namespace/configurator/mod.rs:129]: why not use `async_trait`?

File: libsql-server/src/http/admin/stats.rs [hunk 3]
```diff
@@ -140,10 +140,12 @@ pub(super) async fn handle_stats<C>(
     State(app_state): State<Arc<AppState<C>>>,
     Path(namespace): Path<String>,
 ) -> crate::Result<Json<StatsResponse>> {
+    dbg!();
```
Comment (LucioFranco) [inline libsql-server/src/http/admin/stats.rs]: I am reviewing commit by commit, so not sure if these got removed but wanted to note it.

File: libsql-server/src/lib.rs [hunk 4]
```diff
@@ -623,16 +581,19 @@ where
                 .configure(&mut join_set);
             }
             DatabaseKind::Replica => {
+                dbg!();
```
Comment (LucioFranco) [inline libsql-server/src/lib.rs]: more here

File: libsql-server/src/lib.rs [hunk 5]
```diff
@@ -28,6 +28,7 @@ use auth::Auth;
 use config::{
     AdminApiConfig, DbConfig, HeartbeatConfig, RpcClientConfig, RpcServerConfig, UserApiConfig,
 };
+use futures::future::{pending, ready};
```
Comment (LucioFranco) [inline libsql-server/src/lib.rs:31]: btw these exist now in std https://doc.rust-lang.org/stable/std/future/index.html

**Review by LucioFranco — APPROVED** [review]

File: libsql-server/src/namespace/configurator/mod.rs (hunk 2 above)
Comment (MarinPostma) [reply libsql-server/src/namespace/configurator/mod.rs:129]: no real reason, except that async_trait doesn't play nicely with LSP, and it expends to something quite trivial...

File: libsql-server/src/namespace/configurator/primary.rs (hunk 1 above)
Comment (MarinPostma) [reply libsql-server/src/namespace/configurator/primary.rs:86]: this is not new, just moved over here. if that fails, the namespace won't be loaded, so we assume that there will be a retry. I that keeps failing forever then yes, but that's another problem (like out of disk or some stuff like that)

File: libsql-server/src/lib.rs (hunk 5 above)
Comment (MarinPostma) [reply libsql-server/src/lib.rs:31]: ah right, for some reason RA imports those by default

**Pull request merged**